File: cmd/cli/templates/mobile/ios/App/Runtime/TokenStore.swift
import Foundation
import Security

/// Stored authentication token
public struct AuthToken: Codable, Sendable {
    public let accessToken: String
    public let refreshToken: String?
    public let expiresAt: Date?
    public let tokenType: String

    public var isExpired: Bool {
        guard let expiresAt = expiresAt else { return false }
        return Date() >= expiresAt
    }

    public init(
        accessToken: String,
        refreshToken: String? = nil,
        expiresAt: Date? = nil,
        tokenType: String = "Bearer"
    ) {
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.expiresAt = expiresAt
        self.tokenType = tokenType
    }
}

/// Token storage protocol
public protocol TokenStore: Sendable {
    func getToken() -> AuthToken?
    func setToken(_ token: AuthToken) throws
    func clearToken() throws
    func onTokenChange(_ callback: @escaping @Sendable (AuthToken?) -> Void)
}

/// Keychain-backed token storage
public final class KeychainTokenStore: TokenStore, @unchecked Sendable {
    private let service: String
    private let account: String
    private var observers: [@Sendable (AuthToken?) -> Void] = []
    private let lock = NSLock()

    public init(
        service: String = Bundle.main.bundleIdentifier ?? "com.mizu.app",
        account: String = "auth_token"
    ) {
        self.service = service
        self.account = account
    }

    public func getToken() -> AuthToken? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]

        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        guard status == errSecSuccess,
              let data = result as? Data else {
            return nil
        }

        return try? JSONDecoder().decode(AuthToken.self, from: data)
    }

    public func setToken(_ token: AuthToken) throws {
        let data = try JSONEncoder().encode(token)

        // Delete existing token first
        try? clearToken()

        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]

        let status = SecItemAdd(query as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw KeychainError.unableToSave(status)
        }

        notifyObservers(token)
    }

    public func clearToken() throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account
        ]

        let status = SecItemDelete(query as CFDictionary)
        guard status == errSecSuccess || status == errSecItemNotFound else {
            throw KeychainError.unableToDelete(status)
        }

        notifyObservers(nil)
    }

    public func onTokenChange(_ callback: @escaping @Sendable (AuthToken?) -> Void) {
        lock.withLock {
            observers.append(callback)
        }
    }

    private func notifyObservers(_ token: AuthToken?) {
        let currentObservers = lock.withLock { observers }
        currentObservers.forEach { $0(token) }
    }
}

/// Keychain errors
public enum KeychainError: LocalizedError {
    case unableToSave(OSStatus)
    case unableToDelete(OSStatus)

    public var errorDescription: String? {
        switch self {
        case .unableToSave(let status):
            return "Unable to save to keychain: \(status)"
        case .unableToDelete(let status):
            return "Unable to delete from keychain: \(status)"
        }
    }
}

/// In-memory token store for testing
public final class InMemoryTokenStore: TokenStore, @unchecked Sendable {
    private var token: AuthToken?
    private var observers: [@Sendable (AuthToken?) -> Void] = []
    private let lock = NSLock()

    public init() {}

    public func getToken() -> AuthToken? {
        lock.withLock { token }
    }

    public func setToken(_ token: AuthToken) throws {
        lock.withLock {
            self.token = token
        }
        notifyObservers(token)
    }

    public func clearToken() throws {
        lock.withLock {
            self.token = nil
        }
        notifyObservers(nil)
    }

    public func onTokenChange(_ callback: @escaping @Sendable (AuthToken?) -> Void) {
        lock.withLock {
            observers.append(callback)
        }
    }

    private func notifyObservers(_ token: AuthToken?) {
        let currentObservers = lock.withLock { observers }
        currentObservers.forEach { $0(token) }
    }
}

File: cmd/cli/templates/mobile/reactnative/src/runtime/tokenStore.ts
import * as SecureStore from 'expo-secure-store';

const TOKEN_KEY = 'mizu_auth_token';

export interface AuthToken {
  accessToken: string;
  refreshToken?: string;
  expiresAt?: string;
  tokenType: string;
}

export function createAuthToken(params: {
  accessToken: string;
  refreshToken?: string;
  expiresAt?: Date;
  tokenType?: string;
}): AuthToken {
  return {
    accessToken: params.accessToken,
    refreshToken: params.refreshToken,
    expiresAt: params.expiresAt?.toISOString(),
    tokenType: params.tokenType ?? 'Bearer',
  };
}

export function isTokenExpired(token: AuthToken): boolean {
  if (!token.expiresAt) return false;
  return new Date() >= new Date(token.expiresAt);
}

export type TokenChangeCallback = (token: AuthToken | null) => void;

export interface TokenStore {
  getToken(): Promise<AuthToken | null>;
  setToken(token: AuthToken): Promise<void>;
  clearToken(): Promise<void>;
  onTokenChange(callback: TokenChangeCallback): () => void;
}

/**
 * Secure storage-backed token storage using Expo SecureStore
 */
export class SecureTokenStore implements TokenStore {
  private observers: Set<TokenChangeCallback> = new Set();

  async getToken(): Promise<AuthToken | null> {
    try {
      const json = await SecureStore.getItemAsync(TOKEN_KEY);
      if (!json) return null;
      return JSON.parse(json) as AuthToken;
    } catch {
      return null;
    }
  }

  async setToken(token: AuthToken): Promise<void> {
    await SecureStore.setItemAsync(TOKEN_KEY, JSON.stringify(token));
    this._notifyObservers(token);
  }

  async clearToken(): Promise<void> {
    await SecureStore.deleteItemAsync(TOKEN_KEY);
    this._notifyObservers(null);
  }

  onTokenChange(callback: TokenChangeCallback): () => void {
    this.observers.add(callback);
    return () => this.observers.delete(callback);
  }

  private _notifyObservers(token: AuthToken | null): void {
    this.observers.forEach((callback) => callback(token));
  }
}

/**
 * In-memory token store for testing
 */
export class InMemoryTokenStore implements TokenStore {
  private token: AuthToken | null = null;
  private observers: Set<TokenChangeCallback> = new Set();

  async getToken(): Promise<AuthToken | null> {
    return this.token;
  }

  async setToken(token: AuthToken): Promise<void> {
    this.token = token;
    this._notifyObservers(token);
  }

  async clearToken(): Promise<void> {
    this.token = null;
    this._notifyObservers(null);
  }

  onTokenChange(callback: TokenChangeCallback): () => void {
    this.observers.add(callback);
    return () => this.observers.delete(callback);
  }

  private _notifyObservers(token: AuthToken | null): void {
    this.observers.forEach((callback) => callback(token));
  }
}
